### PR TITLE
Faster final exponentiation for BLS12

### PR DESCRIPTION
**Pull request by Pratyush**
## Description

Implements the faster final exponentiation from https://eprint.iacr.org/2020/875.pdf. In my local benchmarks, this speeds up final exponentiation by ~2%

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (master)
- [x] Linked to Github issue with discussion and accepted design OR have an explanation in the PR that describes this work.
- [x] Wrote unit tests
- [x] Updated relevant documentation in the code
- [x] Added a relevant changelog entry to the `Pending` section in `CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the Github PR explorer

### CODE REVIEW

**Comment by Pratyush** [comment]
cc @cc @yelhousni this is adapted from your code

File: ec/src/models/bls12/mod.rs [hunk 1]
```diff
@@ -145,35 +143,46 @@ impl<P: Bls12Parameters> PairingEngine for Bls12<P> {
...
-            y4 *= &y3;
-            y4.frobenius_map(1);
-            y5 *= &y2;
-            y5.frobenius_map(2);
-            y5 *= &y0;
-            y5 *= &y4;
-            y5 *= &y1;
-            y5
+            // t[1].Expt(&result)
+            let mut y1 = Self::exp_by_x(r);
```
Comment (ValarDragon) [inline ec/src/models/bls12/mod.rs:150]: Lets label what is `x` and `x/2` in both this function, and as a comment on the function within the file

Comment (ValarDragon) [reply ec/src/models/bls12/mod.rs:156]: Thanks for adding it! Didn't realize it was literally the bls12 param

**Comment by Pratyush** [comment]
This is ready for a final review + merge

File: ec/src/models/bls12/mod.rs [hunk 2]
```diff
@@ -145,35 +149,46 @@ impl<P: Bls12Parameters> PairingEngine for Bls12<P> {
...
+            // t[2].InverseUnitary(&result)
+            let mut y2 = r;
+            y2.conjugate();
+            // t[1].Mul(&t[1], &t[2])
+            y1 *= y2;
+            // t[2].Expt(&t[1])
+            y2 = Self::exp_by_x(y1);
```
Comment (mratsim) [inline ec/src/models/bls12/mod.rs:163]: I would be wary if copies of Fp12 elements, in my experience the compiler produce a bench of movaps, movups instructions, you might want an out-of-place pow_by_x

**Review by mratsim — COMMENTED** [review]
I would introduce out-of-place `pow_by_x` functions.

Also I don't see you using the x/2 trick which also exists in the 2016 paper, a cyclotomic is an interesting saving.

For reference this is my implementation: https://github.com/mratsim/constantine/blob/5806cc46387752d4d011ba663eca0766ff4561e4/constantine/pairing/pairing_bls12.nim#L116-L147

File: ec/src/models/bls12/mod.rs [hunk 3]
```diff
@@ -145,35 +149,46 @@ impl<P: Bls12Parameters> PairingEngine for Bls12<P> {
...
+            // t[1].Mul(&t[1], &t[2])
+            y1 *= &y2;
+            // result.Mul(&result, &t[0])
+            r *= y0;
+            // t[0].Expt(&t[1])
+            y0 = Self::exp_by_x(y1);
+            // t[2].Expt(&t[0])
+            y2 = Self::exp_by_x(y0);
+            // t[0].FrobeniusSquare(&t[1])
+            y0 = y1;
```
Comment (mratsim) [inline ec/src/models/bls12/mod.rs:181]: Ditto for the copy here

File: ec/src/models/bls12/mod.rs [hunk 4]
```diff
@@ -145,35 +149,46 @@ impl<P: Bls12Parameters> PairingEngine for Bls12<P> {
...
-            y5 *= &y4;
-            y5 *= &y1;
-            y5
+            // t[1].Expt(&result)
+            let mut y1 = Self::exp_by_x(r);
+            // t[2].InverseUnitary(&result)
+            let mut y2 = r;
+            y2.conjugate();
+            // t[1].Mul(&t[1], &t[2])
+            y1 *= y2;
```
Comment (Pratyush) [inline ec/src/models/bls12/mod.rs:161]:
```suggestion
            y1 *= &y2;
```

File: ec/src/models/bls12/mod.rs [hunk 5]
```diff
@@ -145,35 +149,46 @@ impl<P: Bls12Parameters> PairingEngine for Bls12<P> {
...
+            // t[2].Expt(&t[1])
+            y2 = Self::exp_by_x(y1);
+            // t[1].InverseUnitary(&t[1])
+            y1.conjugate();
+            // t[1].Mul(&t[1], &t[2])
+            y1 *= y2;
```
Comment (Pratyush) [inline ec/src/models/bls12/mod.rs:167]:
```suggestion
            y1 *= &y2;
```

File: ec/src/models/bls12/mod.rs [hunk 6]
```diff
@@ -145,35 +149,46 @@ impl<P: Bls12Parameters> PairingEngine for Bls12<P> {
...
+            // t[2].Expt(&t[1])
+            y2 = Self::exp_by_x(y1);
+            // t[1].Frobenius(&t[1])
+            y1.frobenius_map(1);
+            // t[1].Mul(&t[1], &t[2])
+            y1 *= &y2;
+            // result.Mul(&result, &t[0])
+            r *= y0;
```
Comment (Pratyush) [inline ec/src/models/bls12/mod.rs:175]:
```suggestion
            r *= &y0;
```

File: ec/src/models/bls12/mod.rs [hunk 7]
```diff
@@ -145,35 +149,46 @@ impl<P: Bls12Parameters> PairingEngine for Bls12<P> {
...
+            y0 = y1;
+            y0.frobenius_map(2);
+            // t[1].InverseUnitary(&t[1])
+            y1.conjugate();
+            // t[1].Mul(&t[1], &t[2])
+            y1 *= y2;
+            // t[1].Mul(&t[1], &t[0])
+            y1 *= y0;
+            // result.Mul(&result, &t[1])
+            r *= y1;
```
Comment (Pratyush) [inline ec/src/models/bls12/mod.rs:190]:
```suggestion
            y1 *= &y2;
            // t[1].Mul(&t[1], &t[0])
            y1 *= &y0;
            // result.Mul(&result, &t[1])
            r *= &y1;
```

File: ec/src/models/bls12/mod.rs (hunk 2 above)
Comment (Pratyush) [reply ec/src/models/bls12/mod.rs:163]: To clarify, you mean that `exp_by_x` should take `x` by reference? Or?

File: ec/src/models/bls12/mod.rs (hunk 3 above)
Comment (Pratyush) [reply ec/src/models/bls12/mod.rs:181]: Hm how do I avoid the copy of `y1` here?

**Review by ValarDragon — APPROVED** [review]
This LGTM.

Should we leave todo's in the code / or generally usage of fp12 regarding the copies?

File: ec/src/models/bls12/mod.rs (hunk 3 above)
Comment (mratsim) [reply ec/src/models/bls12/mod.rs:181]: This is how I did it

```Nim
  # (x²+p²−1)
  v2.pow_x(v0, invert = false)
  v1.pow_x(v2, invert = false) # v1 = f^((x-1)².(x+p).x²)
  v2.frobenius_map(v0, 2)      # v2 = f^((x-1)².(x+p).p²)
  v0.cyclotomic_inv()          # v0 = f^((x-1)².(x+p).-1)
  v0 *= v1                     # v0 = f^((x-1)².(x+p).(x²-1))
  v0 *= v2                     # v0 = f^((x-1)².(x+p).(x²+p²-1))

  # (x−1)².(x+p).(x²+p²−1) + 3
  f *= v0
```

File: ec/src/models/bls12/mod.rs (hunk 2 above)
Comment (mratsim) [reply ec/src/models/bls12/mod.rs:163]: I mean mutate `y2` in-place

**Comment by Pratyush** [comment]
Pushed some changes which optimize the copying behaviour; this improves the performance by another 2%, for an overall 4% improvement.

**Pull request merged**